### ISSUE TITLE
[Discussion] Should RandomGenerator be public or internal?

### ISSUE THREAD

**Issue by Doug-Murphy** [nbuilder/nbuilder#112]
It seems to me that the RandomGenerator class maybe shouldn't be public, but rather it should be internal since it seems to be (at least almost) a complete duplicate of GetRandom. GetRandom uses this class exclusively to do all of its random generation, so exposing RandomGenerator could become confusing to certain users.

Is there something I'm not seeing? Should things be kept as-is to avoid any breaking changes? Should this be changed to internal and released as version 7.0.0?

**Comment by PureKrome** [comment]
If it's a duplicate, then obsolete or internal. I would prefer obsolete because why have 2x versions of the same code?

yes, those 2 options are breaking changes, etc. which is fine with a major semver release IMO.

**Comment by Doug-Murphy** [comment]
@PureKrome I'll have to do some more looking into this and create a list of shared functionality. Initial glance at least revealed a significant amount of shared functionality.

My current preference is that whatever exists in `RandomGenerator` but not in `GetRandom` should be ported over, then `RandomGenerator` deleted. I would be in favor of this breaking change and releasing a new major version as per semver.

I'll do further looking at it this week and post back with some updated findings and maybe an updated proposal.

**Comment by PureKrome** [comment]
![](https://media.giphy.com/media/xT9IgrNBmDXi6Z0YoM/giphy.gif)

**Comment by Doug-Murphy** [comment]
It looks like `RandomGenerator` is used heavily by `GetRandom` (as stated above) since `GetRandom` is more of a wrapper around `RandomGenerator`. However, I can see a use-case for wanting to keep `RandomGenerator` exposed publicly since it's a bit "lower level" in its use. I personally still think having it as `internal` makes sense, but I can see why some users may want the lower level access that `RandomGenerator` provides, even if it's not necessarily better in my opinion. :)

It is possible there is a use-case that `RandomGenerator` provides that `GetRandom` does not.

If there are no further thoughts this weekend, I'll just close this discussion.

**Comment by PureKrome** [comment]
I'd still vote to simplifying all of this.

There could easily be a roadmap for this -> obsolete it for a major semver version and see if there are any complaints. If there are, re-evaluate. Otherwise, we're good to go.

I personally prefer simplified API's/Libraries instead of too many options which could confuse consumers/developers to which way to go/use.

**Comment by Doug-Murphy** [comment]
I completely agree with you on this. I think it's probably better overall to consolidate these into one interface (GetRandom static class in my opinion). I just error on the side of not removing functionality from users who may be expecting it. But so long as an alternative is provided for them... it should be ok.

Do you want to take a stab at consolidating any missing functionality from RandomGenerator into GetRandom and marking RandomGenerator as Obsolete? Then we make it internal (since GetRandom uses it) for a next major release.

**Comment by PureKrome** [comment]
Kewl - nice to see someone else thinking similar to me :) I'm also not a maintainer of this library, just someone who's been using it for _ages_ so I sorta interested in seeing where things go with it.

So IMO .. yep, give this a _first pass_ go. If it's going to take more than an hour or 2, I'd wait until @garethdown44 or @crmckenzie (the 2 repo maintainers?) throws his thoughts in here .. considering he'll be the person to accept the PR or not?

SIDE NOTE 1: Not sure how to find out who are the official list of repo owners/maintainers
SIDE NOTE 2: linked to #46

**Comment by crmckenzie** [comment]
Official maintainers right now is me. Gareth is a victim of love, marriage, and parenthood so doesn't meaningfully contribute anymore. I'm due for a another round of PR review so I'll get on that soon.

Regarding this one, I can see how RandomGenerator could be useful in isolation, but I don't see a need for multiple implementations. I'm definitely in favor of consolidating into a single instance--at face value anyway.

**Comment by Doug-Murphy** [comment]
Good timing for me to chime in seeing as this was active literally 1 minute ago. :D

I just looked at RandomGenerator and did a feature comparison with GetRandom. I can confirm that GetRandom includes 100% of RandomGenerator. I will open a quick PR to just mark RandomGenerator as [Obsolete] so that there is time for it to sit a bit pending marking it internal (or removing it?) for the next major release.

**Comment by Doug-Murphy** [comment]
@crmckenzie Are you available to address the lingering PRs? :)

**Comment by PureKrome** [comment]
If there's anyone here reading this .. I'm actually hijacking this issue and reverting this action.

Since my initial thoughts waaaaay back, I've really started to move away from STATIC's - especially when they contain state. With this `GetRandom` static class, it actually does have state -> the random generator.

Instead, i've 
- marked `GetRandom.cs` as `Obsolete`, 
- added all the static methods (from `GetRandom`) to `IRandomGenerator`
- made all the static methods in `GetRandom` just delegate to _an internal instance_ of `RandomGenerator`
- **added a `Default` lazy singleton** to `RandomGenerator` (similar to how other classes have `.Default` property like `CancellationToken.Default`, etc)
- Made the `GetRandom` static class internally call `RandomGenerator.Default` 

so this way
- `GetRandom` is backwards compat (not a breaking change)
- `RandomGenerator` is updated (not a breaking change)

and we can nuke `GetRandom` later.

changes were made in this commit: a4ea69c925264a3d6f2e244617871a8b024ec451
